packages/troika-three-text/src/Typesetter.js

```diff
@@ -517,13 +521,17 @@ export function createTypesetter(resolveFonts, bidi) {
                 caretPositions[charIndex * 4 + 2] = line.baseline + fontData.caretBottom + anchorYOffset //common bottom y
                 caretPositions[charIndex * 4 + 3] = line.baseline + fontData.caretTop + anchorYOffset //common top y
 
-                // If we skipped any chars from the previous glyph (due to ligature subs), fill in caret
-                // positions for those missing char indices; currently this uses a best-guess by dividing
+                // If we skipped any chars from the previous glyph (due to ligature subs/surrogates), fill in caret
+                // positions for those missing char indices; currently ligatures uses a best-guess by dividing
                 // the ligature's width evenly. In the future we may try to use the font's LigatureCaretList
                 // table to get better interior caret positions.
-                const ligCount = charIndex - prevCharIndex
-                if (ligCount > 1) {
-                  fillLigatureCaretPositions(caretPositions, prevCharIndex, ligCount)
+                const charCount = charIndex - prevCharIndex
+                if (charCount > 1) {
+                  if (isSurrogate(text.slice(prevCharIndex, charIndex))) {
+                    fillSurrogateCaretPositions(caretPositions, prevCharIndex, charCount)
+                  } else {
+                    fillLigatureCaretPositions(caretPositions, prevCharIndex, charCount)
+                  }
```

Comment:
I feel like this may not be a strict if/else situation ... you could potentially have ligatures formed from codepoints that are surrogates. Maybe a better approach would be to modify `fillLigatureCaretPositions` to check for surrogates stepwise within its loop?

Reply:
You're correct, took a deeper look into things and I didn't realize the tons of other characters that are considered surrogate pairs that aren't emojis and could very likely be part of ligature substitutions. Thanks for calling this out